scripts/Muon/GUI/Common/grouping_tab_widget/grouping_tab_widget_presenter.py

```diff
@@ -0,0 +1,217 @@
+
+        # multi-threading
+        self.thread_manager = WorkHandler()
+
+        # monitors for loaded data changing
+        self.loadObserver = GroupingTabPresenter.LoadObserver(self)
+        self.instrumentObserver = GroupingTabPresenter.InstrumentObserver(self)
+
+        # notifiers
+        self.groupingNotifier = GroupingTabPresenter.GroupingNotifier(self)
+        self.grouping_table_widget.on_data_changed(self.group_table_changed)
+        self.pairing_table_widget.on_data_changed(self.pair_table_changed)
+
+        self.guessAlphaObserver = GroupingTabPresenter.GuessAlphaObserver(self)
+        self.pairing_table_widget.guessAlphaNotifier.add_subscriber(self.guessAlphaObserver)
+
+    def show(self):
+        self._view.show()
+
+    def text_for_description(self):
+        """
+        Generate the text for the description edit at the top of the widget.
+        """
+        instrument = self._model.instrument
+        n_detectors = self._model.num_detectors
+        main_field = self._model.main_field_direction
+        text = "{} , {} detectors, main field : {} to muon polarization".format(
+            instrument, n_detectors, main_field)
+        return text
+
+    def update_description_text(self, description_text=''):
+        if not description_text:
+            description_text = self.text_for_description()
+        self._view.set_description_text(description_text)
+
+    def add_pair_from_grouping_table(self, group_name1, group_name2):
+        """
+        If user requests to add a pair from the grouping table.
+        """
+        pair = self._model.construct_empty_pair_with_group_names(group_name1, group_name2)
+        self._model.add_pair(pair)
+        self.pairing_table_widget.update_view_from_model()
+
+    def handle_guess_alpha(self, pair_name, group1_name, group2_name):
+        """
+        Calculate alpha for the pair for which "Guess Alpha" button was clicked.
+        """
+        ws1 = self._model.get_group_workspace(group1_name)
+        ws2 = self._model.get_group_workspace(group2_name)
+
+        ws = algorithm_utils.run_AppendSpectra(ws1, ws2)
+
+        new_alpha = algorithm_utils.run_AlphaCalc({"InputWorkspace": ws,
+                                                   "ForwardSpectra": [0],
+                                                   "BackwardSpectra": [1]})
+        self._model.update_pair_alpha(pair_name, new_alpha)
+        self.pairing_table_widget.update_view_from_model()
+
+        self.groupingNotifier.notify_subscribers()
+        self.handle_update_all_clicked()
+
+    def handle_load_grouping_from_file(self):
+        # Only XML format
+        file_filter = file_utils.filter_for_extensions(["xml"])
+        filename = self._view.show_file_browser_and_return_selection(file_filter, [""])
+
+        groups, pairs, description = xml_utils.load_grouping_from_XML(filename)
+
+        try:
+            self._model.clear()
+            for group in groups:
```

Comment:
could we have the try catch in the for loops? I know its not as neat, but it would mean that if say your 2nd of 4 groups is wrong you would still load the other 3 and the pairs instead of just the first one.

Reply:
Do we want to load the pairs if the groups they depend on are corrupted though. I originally did it this way so the pairs would not be loaded if the groups were wrong.

Reply:
say we have three groups (1, 2, 3) and group 2 is corrupt. Then I would still like 1 and 3 in the ADS. If I have two pairs (4, 5) and only pair 4 depended on group 2, I would still like to see pair 5.